compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp

```diff
@@ -926,6 +927,28 @@ static LogicalResult setWarpReductionConfig(func::FuncOp entryPoint,
   if ((groupSize / subgroupSize) > subgroupSize)
     return failure();
 
+  // With just one subgroup per workgroup, make each subgroup do more work and
+  // process a few reductions along the last parallel dimension.
+  // TODO: We should also check that this will result in data reuse for at least
+  //       one argument.
+  // TODO: This is experimental and rocm-only for now.
+  if (isRocmTarget(entryPoint)) {
+    if (*parallelSize && !parallelDims.empty() && groupSize == subgroupSize) {
+      int reductionsPerWorkgroup = 1;
+      int maxParallelFactor = 4; // Keeping this conservative for now.
+      int64_t lastParallelBound = bounds[parallelDims.back()];
+      if (!ShapedType::isDynamic(lastParallelBound) &&
+          (lastParallelBound % maxParallelFactor == 0) &&
+          lastParallelBound > maxParallelFactor) {
+        for (int candidate = reductionsPerWorkgroup;
+             candidate <= maxParallelFactor; candidate *= 2) {
+          reductionsPerWorkgroup = candidate;
```

Comment:
So does this always starts with 1, and ends at 4? Since it doesn't look like there are other checks or heuristics to exit the loop. If so, couldn't we just set it to 4 all the time?

Reply:
It can be 2 if the number of rows is not divisible by 4. I saw some dispatches that were not divisible by 8 which was one of the max sizes I tried, so I'd rather keep this logic.

Reply:
oh I think I accidentally dropped this check... good catch